src/main/scala/io/iohk/ethereum/forkid/ForkIdValidator.scala

```diff
@@ -0,0 +1,129 @@
+package io.iohk.ethereum.forkid
+
+import akka.util.ByteString
+import cats.Monad
+import cats.data.EitherT
+import cats.implicits._
+import io.iohk.ethereum.utils.BigIntExtensionMethods._
+import io.iohk.ethereum.utils.BlockchainConfig
+import io.iohk.ethereum.utils.ByteUtils._
+import monix.eval.Task
+import org.typelevel.log4cats.Logger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+import java.util.zip.CRC32
+
+sealed trait ForkIdValidationResult
+case object Connect extends ForkIdValidationResult
+case object ErrRemoteStale extends ForkIdValidationResult
+case object ErrLocalIncompatibleOrStale extends ForkIdValidationResult
+
+object ForkIdValidator {
+
+  implicit val unsafeLogger = Slf4jLogger.getLogger[Task]
+
+  val maxUInt64 = (BigInt(0x7fffffffffffffffL) << 1) + 1 // scalastyle:ignore magic.number
+
+  def validatePeer[F[_]: Monad: Logger](
+      genesisHash: ByteString,
+      config: BlockchainConfig
+  )(currentHeight: BigInt, remoteForkId: ForkId): F[ForkIdValidationResult] = {
+    val forks = ForkId.gatherForks(config)
+    validatePeer[F](genesisHash, forks)(currentHeight, remoteForkId)
+  }
+
+  def validatePeer[F[_]: Monad: Logger](
+      genesisHash: ByteString,
+      forks: List[BigInt]
+  )(currentHeight: BigInt, remoteId: ForkId): F[ForkIdValidationResult] = {
+    val checksums: Vector[BigInt] = calculateCheckchecksums(genesisHash, forks)
+
+    // find the first unpassed fork and it's index
+    val (unpassedFork, i) = (forks :+ maxUInt64).zipWithIndex.find { case (fork, _) => currentHeight < fork }.get
+
+    // The checks are left biased -> whenever a result is found we need to short circuit
+    val validate: F[Option[ForkIdValidationResult]] = (for {
+      _ <- EitherT.liftF(Logger[F].debug(s"Before checkMatchingHashes"))
+      matching <- EitherT.fromEither[F](
+        checkMatchingHashes(checksums, remoteId, currentHeight, i).toLeft("hashes didn't match")
+      )
+      _ <- EitherT.liftF(Logger[F].debug(s"checkMatchingHashes result: $matching"))
+      _ <- EitherT.liftF(Logger[F].debug(s"Before checkSubset"))
+      sub <- EitherT.fromEither[F](checkSubset(checksums, forks, remoteId, i).toLeft("not in subset"))
+      _ <- EitherT.liftF(Logger[F].debug(s"checkSubset result: $sub"))
+      _ <- EitherT.liftF(Logger[F].debug(s"Before checkSuperset"))
+      sup <- EitherT.fromEither[F](checkSuperset(checksums, remoteId, i).toLeft("not in superset"))
+      _ <- EitherT.liftF(Logger[F].debug(s"checkSuperset result: $sup"))
+      _ <- EitherT.liftF(Logger[F].debug(s"No check succeeded"))
+      res <- EitherT.fromEither[F](Either.left[ForkIdValidationResult, Unit](ErrLocalIncompatibleOrStale))
```

Comment:
there's a lot of boilerplate simply to get a lifted debug. perhaps a helper?
`private def liftedDebug[F[_]: Monad: Logger](s: => String) = EitherT.liftF(Logger[F].debug(s))`

```
    val validate: F[Option[ForkIdValidationResult]] = (for {
      _ <- liftedDebug[F](s"Before checkMatchingHashes")
      matching <- EitherT.fromEither[F](
        checkMatchingHashes(checksums, remoteId, currentHeight, i).toLeft("hashes didn't match")
      )
      _ <- liftedDebug[F](s"checkMatchingHashes result: $matching")
      _ <- liftedDebug[F](s"Before checkSubset")
      sub <- EitherT.fromEither[F](checkSubset(checksums, forks, remoteId, i).toLeft("not in subset"))
      _ <- liftedDebug[F](s"checkSubset result: $sub")
      _ <- liftedDebug[F](s"Before checkSuperset")
      sup <- EitherT.fromEither[F](checkSuperset(checksums, remoteId, i).toLeft("not in superset"))
      _ <- liftedDebug[F](s"checkSuperset result: $sup")
      _ <- liftedDebug[F](s"No check succeeded")
```

Reply:
I imported `EitherT._` and use the lifting functions without the explicit companion object